Pseudonomous identity and use the new jwt service endpoint (with delayed event delegation)

toger5 opened the pull request:
This also does all the compatibility work. When to use which endpoint to authenticate agains a jwt service.

toger5 committed:
still with broken tests...

toger5 committed:
small refactor to make it testable.

toger5 committed:
Merge branch 'livekit' into toger5/pseudonomous-identities

fkwp committed:
switch synapse docker image to one with sticky event support

fkwp committed:
enable sticky events

toger5 committed:
Merge branch 'livekit' into toger5/pseudonomous-identities

toger5 committed:
Make use of the new jwt service endpoint (with delayed event delegation)

This also does all the compatibility work. When to use which endpoint to
authenticate agains a jwt service.

toger5 commented:
Looking at the changes it might make more sense to go with a retry system for the jwt so we do not have to pass aroudn the useMatrix2 information with the transport.

toger5 committed:
use correct js-sdk

toger5 committed:
temp refactored membership rtcidentity

toger5 committed:
use the js-sdk where the hashed rtcSessionIdeintity is already part of
the event (no need to compute it in the encryption manager)

toger5 committed:
Merge branch 'livekit' into toger5/delayed-event-delegation

toger5 committed:
cleanup based on new js-sdk impl

BillCarsonFr commented on src/livekit/openIDSFU.ts:
```diff
@@ -52,26 +62,35 @@ export async function getSFUConfigWithOpenID(
-  if (matrix2jwt) {
+  try {
+    // we do not want to try the old endpoint, since we are not sending the new matrix2.0 sticky events (no hashed identity in the event)
+    if (forceOldJwtEndpoint) throw new Error("Force old jwt endpoint");
```
I am not really a fan of using an exception as a conditional control flow. 
I'd prefer that this is formatted as a normal conditional

BillCarsonFr commented on src/state/CallViewModel/localMember/LocalTransport.ts:
```diff
@@ -62,6 +61,11 @@ interface Props {
+ * @prop useOldJwtEndpoint$ Whether to set forceOldJwtEndpoint the use the old JWT endpoint.
+ * This is used when the connection manager needs to know if it has to use the legacy endpoint which implies a string concatenated rtcBackendIdentity.
+ * (which is expected for non sticky event based rtc member events)
+ * @returns Behavior<(LivekitTransport & { forceOldJwtEndpoint: boolean }) | null> The `forceOldJwtEndpoint` field is added to let the connection EncryptionManager
```
The @returns doc is outdated

BillCarsonFr commented on src/state/CallViewModel/localMember/LocalTransport.ts:
```diff
@@ -142,6 +144,8 @@ const FOCI_WK_KEY = "org.matrix.msc4143.rtc_foci";
  *
  * @param client
  * @param roomId
+ * @param useMatrix2 This implies using the matrix2 jwt endpoint (including delayed event delegation of the jwt token)
```
outdated doc. 
I think we miss some js-doc CI? It owuld help catching those

toger5 replied:
there are added now right?

toger5 replied:
doc string does not mention this anymore

toger5 committed:
Merge branch 'livekit' into toger5/delayed-event-delegation

toger5 committed:
fixup merge

toger5 committed:
simplify localTransport

toger5 committed:
use latest js-sdk

toger5 committed:
Merge branch 'livekit' into toger5/delayed-event-delegation

toger5 committed:
self review

toger5 committed:
fix js-doc

toger5 committed:
fix tests and remove duplicated mocks.

toger5 committed:
add tests for open id delay fallback

toger5 committed:
fix lint

toger5 committed:
dont use throw for logic.

BillCarsonFr commented on src/livekit/MatrixAudioRenderer.tsx:
```diff
@@ -32,7 +31,7 @@ export interface MatrixAudioRendererProps {
    * that are not expected to be in the rtc session (local user is excluded).
    */
-  validIdentities: ParticipantId[];
+  validIdentities: string[];
```
This is the rtcBackendIdentity now I guess. 
Would it be worth creating an alias for that in the js-sdK?

toger5 replied:
I dont think so. There are no string semantics to that identity... The names should already imply what it is?

But If we have a PR that introduces such a thing I would not object.

toger5 committed:
review (docstrings) and remove unused artifacts.

BillCarsonFr commented on src/livekit/openIDSFU.ts:
```diff
@@ -84,16 +101,44 @@ export async function getSFUConfigWithOpenID(
+        delayId,
+      );
+      logger?.info(`Got JWT from call's active focus URL.`);
+    } catch (e) {
```
Shouldn't we be a bit more specific regarding the error here? What if it is just a transiant network error? Can we check for the exact incompatible error instead?

toger5 replied:
Added a custom error type. So we can only try with the new endpoint if we fail because the `/get_token` endpoint is not found.

BillCarsonFr commented on src/livekit/openIDSFU.ts:
```diff
@@ -134,3 +182,53 @@ async function getLiveKitJWT(
+  }
+
+  try {
+    const res = await fetch(livekitServiceURL + "/get_token", {
```
We might want to use something like `doNetworkOperationWithRetry` to have some retry for transiant errors.
And in the catch we should look out for the specific error when the modern endpoint is not supported (what is the error? 404 or something else).
We should then throw a custom UnsupportedEndPointError that can be handled by the caller.

BillCarsonFr commented on src/state/CallViewModel/remoteMembers/MatrixLivekitMembers.ts:
```diff
@@ -93,63 +92,56 @@ export function createMatrixLivekitMembers$({
 
             yield {
-              keys: [participantId, membership.userId],
+              keys: [membership.userId, membership.deviceId],
```
I wonder why the `identity` is not part of the key? or maybe the only key?

toger5 replied:
I think we can either only do it based on the rtcBackendIdentity or userId+deviceId otherwise its redundant?

BillCarsonFr commented on dev-backend-docker-compose.yml:
```diff
@@ -88,7 +88,7 @@ services:
   synapse:
     hostname: homeserver
-    image: docker.io/matrixdotorg/synapse:latest
+    image: ghcr.io/element-hq/synapse:pr-18968-dcb7678281bc02d4551043a6338fe5b7e6aa47ce
```
Why do we commit that?

toger5 replied:
Good question. It is the newest version with sticky events. So if we want to create playwrights with sticky events we would need it. But we do not so this is just to allow testing it locally with `yarn backend`
Patching the synapse used for that can also be postponed.

toger5 replied:
I think it makes sense to keep it as the hash for now until we have a stable version (like we do it with the js-sdk branches)

But a playwright test should be added that uses sticky events and relies on that synapse version.

BillCarsonFr reviewed (commented):
Thx, we are getting close 😅
 I have a few open questions

toger5 committed:
cleanup an rename compatibility mode

toger5 committed:
fix up tests

toger5 committed:
fix test snapshot

toger5 committed:
add retries and be more specific when we fall back to legacy endpoint

toger5 committed:
add retries inside the `getLiveKitJWTWithDelayDelegation` and
`getLiveKitJWT` functions.

toger5 committed:
fix tests

toger5 commented:
ugh broke playwright. Will dive into that tomorrow.

toger5 committed:
temp

robintown commented on src/state/CallViewModel/remoteMembers/MatrixLivekitMembers.ts:
```diff
@@ -93,63 +92,56 @@ export function createMatrixLivekitMembers$({
-        ]) =>
-          new Epoch([membershipsWithTransports, managerData] as const, epoch),
-      ),
+      map(([x, y]) => new Epoch([x.value, y.value] as const, x.epoch)),
```
```suggestion
      map(([ms, data]) => new Epoch([ms.value, data.value] as const, ms.epoch)),
```

I like a little name hint…

robintown commented on src/settings/settings.ts:
```diff
@@ -126,7 +126,13 @@ export const alwaysShowIphoneEarpiece = new Setting<boolean>(
 export enum MatrixRTCMode {
   Legacy = "legacy",
-  Compatibil = "compatibil",
+  Compatibility = "compatibility",
+  /** This implies using
+   *  - sticky events
+   *  - hashed RTC backend identity
+   *  - the new endpoint for the jwt token on the local membership (remote memberships will always try the new jwt endpoint first -> then the legacy one)
+   *  - use the hashed identity for the local membership
+   */
```
We need to be mindful in our dogfooding that if any of us have set our client to the compatibility mode previously, you may need to reset it manually (as EC will no longer recognize the old string value)

toger5 replied:
indeed... Lets just keep in in our heads and remember toggling the value in case sth fails.

robintown commented on src/state/CallViewModel/CallViewModel.ts:
```diff
@@ -716,13 +736,13 @@ export function createCallViewModel$(
-            const participantId = `${userId}:${membership$.value.deviceId}`;
-            if (participantId === localParticipantId) continue;
+          } of matrixLivekitMembers.value) {
+            const userMediaId = `${userId}:${membership$.value.deviceId}`;
```
There's another instance of `participantId` just below this in the `factory` callback that should be renamed

robintown commented on src/state/CallViewModel/CallViewModel.ts:
```diff
@@ -685,22 +705,22 @@ export function createCallViewModel$(
-            if (localParticipantId) {
+            localUserMediaId = `${userId}:${membership$.value.deviceId}`; // should be membership$.value.membershipID which is not optional
+
+            if (localUserMediaId) {
```
:thinking: This branch will always be taken, since `localUserMediaId` is necessarily truthy. Don't think we need it.

robintown commented on src/state/CallViewModel/CallViewModel.ts:
```diff
@@ -415,11 +420,38 @@ export function createCallViewModel$(
+    deviceId,
+    // This will eventually become the salt for the hash endpoint.
+    // For now we keep it as the user+device string since it is expected by non matrix matrixRTCMode === Legacy.
+    memberId: `${userId}:${deviceId}`,
```
Hmm, I wonder if we should call this change something other than 'pseudonymous identity' then since there are effectively no pseudonymity guarantees yet. Maybe 'support arbitrary identities' or 'prepare for pseudonymous identities' has less risk of misleading?

toger5 replied:
I think the comment is not even true. memberId is not used at all I think.

toger5 replied:
I checked and we really only consume this if we use the sticky event membership manager.
So we can just make it a uuid now already!

robintown replied:
I'm checking this myself now as well

robintown replied:
Yes, that appears to be the case. Let's use a UUID!

toger5 committed:
fix connection recreation which breaks EC lk connection

toger5 committed:
Refactor how we aquire the jwt token for the local user. (only fetch it
once)

The local jwt token needs to be aquired via the right endpoint. The
endpoint defines how our rtcBackendIdentity is computed. Based on us
using sticky events or state events we also need to use the right
endpoint. This cannot be done generically in the connection manager. The
jwt token now is computed in the localTransport and the resolved sfu
config is passed to the connection manager.

Add JWT endpoint version and SFU config support Pin matrix-js-sdk to a
specific commit and update dev auth image tag. Propagate SFU config and
JWT endpoint choice through local transport, ConnectionManager and
Connection; add JwtEndpointVersion enum and LocalTransportWithSFUConfig
type. Add NO_MATRIX_2 auth error and locale string, thread
rtcBackendIdentity through UI props, and include related test, CSS and
minor imports updates

toger5 committed:
fix js-sdk dependency

BillCarsonFr commented on src/livekit/openIDSFU.ts:
```diff
@@ -113,15 +120,21 @@ export async function getSFUConfigWithOpenID(
-  if (forceOldJwtEndpoint === false) {
+  if (
+    // we do not force anything. Try with new first (remote connections)
+    !opts?.forceJwtEndpoint ||
```
I find this a bit hard to read `!opts?` can it be simplified

BillCarsonFr commented on src/livekit/openIDSFU.ts:
```diff
@@ -113,15 +120,21 @@ export async function getSFUConfigWithOpenID(
+    // we do not force anything. Try with new first (remote connections)
+    !opts?.forceJwtEndpoint ||
+    // we do force the matrix2.0 endpoint
+    (opts?.forceJwtEndpoint &&
```
isn't `opts?.forceJwtEndpoint === JwtEndpointVersion.Matrix_2_0` enough?

toger5 replied:
`opts.forceJwtEndpoint = undefined` would resolve to false then.
(with this it will be true) right?

toger5 replied:
I have an idea how this can be done in a way that si easier to read.

BillCarsonFr commented on src/livekit/openIDSFU.ts:
```diff
@@ -235,7 +252,7 @@ export async function getLiveKitJWTWithDelayDelegation(
 
   let bodyDalayParts = {};
   // Also check for empty string
-  if (delayId && delayEndpointBaseUrl && false) {
```
oh wow what was that

toger5 replied:
The jwt was returning `invalid json` when not using the most recent version.

toger5 replied:
So this was for testing

toger5 committed:
prettier

BillCarsonFr commented on locales/en/app.json:
```diff
@@ -116,6 +116,7 @@
     "matrix_rtc_transport_missing": "The server is not configured to work with {{brand}}. Please contact your server admin (Domain: {{domain}}, Error Code: {{ errorCode }}).",
     "membership_manager": "Membership Manager Error",
     "membership_manager_description": "The Membership Manager had to shut down. This is caused by many consequtive failed network requests.",
+    "no_matrix_2_authorization_service": "Your authorization service for you media server (SFU) is not on the newest version",
```
I can't find it on localazy

toger5 replied:
probably will only appear once we merge?

BillCarsonFr replied:
> probably will only appear once we merge?

We have to do that manually no? First add on localazy then download the files from localazy? no?

robintown replied:
Our English strings are automatically uploaded to Localazy by a GitHub Action on every push to `livekit`: https://github.com/element-hq/element-call/actions/workflows/translations-upload.yaml The file on `livekit` is the source of truth.

BillCarsonFr reviewed (approved):
I'd be ok to merge and continue with follow up PR as long as CI happy and other review request are adressed

toger5 committed:
fix tests

toger5 committed:
use actual uuid for `member.id`

review with some renaming and removal of unnecassary if

toger5 committed:
review valere (more readable force new endpoint logic)

robintown commented on src/state/CallViewModel/CallViewModel.ts:
```diff
@@ -47,6 +47,7 @@ import {
 } from "matrix-js-sdk/lib/matrixrtc";
 import { type IWidgetApiRequest } from "matrix-widget-api";
 import { type CallMembershipIdentityParts } from "matrix-js-sdk/lib/matrixrtc/EncryptionManager";
+import { v4 as uuidv4 } from "uuid";
```
We should add uuid to our dependencies if we are actually using it outside matrix-js-sdk. (Though the built-in [`crypto.randomUUID`](https://developer.mozilla.org/en-US/docs/Web/API/Crypto/randomUUID) function would work just as well here)

toger5 replied:
This caused the following error:
`"randomUUID" is not exported by "__vite-browser-external`

I opted for the dedicated `yarn add -D uuid`

robintown commented on src/state/CallViewModel/CallViewModel.ts:
```diff
@@ -685,29 +714,28 @@ export function createCallViewModel$(
+                  userId,
+                  participant satisfies TaggedParticipant as TaggedParticipant, // Widen the type safely
+                  connection$,
+                  membership$.value,
```
This will be what's causing the issue where the tile reappears after unmuting the camera. `membership$.value` is not a stable key and more to the point, we should design UserMedia to not be tied to a static CallMembership object

toger5 replied:
should we fix this as part of that PR?

toger5 replied:
> to a static CallMembership object

A user media object is a tile and they are based on the callMemberships. Why do we want them to be non static?

I think we even can think of it as the membership being the only key we need for it?

toger5 replied:
I updated this PR to now only pass the `rtcBackendIdenitity`

robintown replied:
> A user media object is a tile and they are based on the callMemberships. Why do we want them to be non static?

Because a user might update their membership during a call to change the intent, for example. Your fix is good.

toger5 committed:
improve rtcBackendIdentity debugging

toger5 committed:
lint

toger5 committed:
bump js-sdk

toger5 committed:
fix tile flicker on video mute/unmute

(only pass rtcBackendIdenitity instead of full rtcMember)

toger5 committed:
fix tests

toger5 committed:
add uuid in our own package.json
